pkg/machine/machineset.go

```diff
@@ -0,0 +1,524 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// MachineSetBuilder provides a struct for MachineSet object from the cluster and a MachineSet definition.
+type MachineSetBuilder struct {
+	// MachineSetBuilder definition. Used to create
+	// MachineSet object with minimum set of required elements.
+	Definition *machinev1beta1.MachineSet
+	// Created MachineSetBuilder object on the cluster.
+	Object *machinev1beta1.MachineSet
+	// api client to interact with the cluster.
+	apiClient *clients.Settings
+	// errorMsg is processed before MachineSetBuilder object is created.
+	errorMsg string
+	// string to store the public cloud
+	publicCloud string
+}
+
+// NewMachineSetBuilderFromCopy returns an MachineSetBuilder struct from a copied MachineSet.
+func NewMachineSetBuilderFromCopy(
+	apiClient *clients.Settings,
+	nsName,
+	instanceType string,
+	replicas int32) *MachineSetBuilder {
+	glog.V(100).Infof(
+		"Initializing new MachineSetBuilder structure from copied MachineSet with the following params: "+
+			"namespace: %s, instanceType: %s and replicas: %v", nsName, instanceType, replicas)
+
+	newMachineSet, err := CreateNewWorkerMachineSetFromCopy(apiClient, nsName, instanceType, replicas)
+
+	builder := MachineSetBuilder{
+		apiClient:  apiClient,
+		Definition: newMachineSet,
+	}
+
+	if err != nil {
+		glog.V(100).Infof(
+			"Error initializing MachineSet from copy: %s", err.Error())
+
+		builder.errorMsg = fmt.Sprintf("Error initializing MachineSet from copy: %s",
+			err.Error())
+	}
+
+	err = builder.GetPublicCloudKind()
+
+	if err != nil {
+		builder.errorMsg = fmt.Sprintf("error getting the public cloud kind: %v", err.Error())
+	}
+
+	glog.V(100).Infof("Updating copied MachineSet provider instanceType to: %s", instanceType)
+
+	err = builder.ChangeCloudProviderInstanceType(instanceType)
+
+	if err != nil {
+		builder.errorMsg = fmt.Sprintf("error changing the instanceType: %v", err.Error())
+	}
+
+	if nsName == "" {
+		glog.V(100).Infof("The Namespace of the MachineSet is empty")
+
+		builder.errorMsg = "MachineSet 'nsName' cannot be empty"
+	}
+
+	if instanceType == "" {
+		glog.V(100).Infof("The instanceType of the MachineSet is empty")
+
+		builder.errorMsg = "MachineSet 'instanceType' cannot be empty"
+	}
+
+	if replicas == 0 {
+		glog.V(100).Infof("The replicas of the MachineSet is zero")
+
+		builder.errorMsg = "MachineSet 'replicas' cannot be zero"
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The MachineSet object definition is nil")
+
+		builder.errorMsg = "MachineSet 'Object.Definition' is nil"
+	}
+
+	return &builder
+}
+
+// GetPublicCloudKind determines the public cloud kind and stores it in the builder struct.
+func (builder *MachineSetBuilder) GetPublicCloudKind() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Determining the public cloud kind")
+
+	providerSpecMap, err := GetMapFromProviderSpec(*builder.Definition)
```

Comment:
I would recommend moving the code from GetMapFromProviderSpec into this method. It is difficult to understand the code when needing to jump between multiple methods/functions

Reply:
Also, it is only used in this one circumstance. We want to discourage anyone from using this over using typed structs

Reply:
Done